libs/num/numeric.go

```diff
@@ -0,0 +1,108 @@
+package num
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+)
+
+type Numeric struct {
+	asUint    *Uint
+	asDecimal *Decimal
+}
+
+func (n *Numeric) Clone() *Numeric {
+	if n.asUint != nil {
+		nn := n
+		return nn
+	}
+
+	nn := &Numeric{}
+
+	if n.asDecimal != nil {
+		decimal := *n.asDecimal
+		nn.asDecimal = &decimal
+	}
+	return nn
+}
+
+func (n *Numeric) String() string {
+	if n.asUint != nil {
+		return n.asUint.String()
+	}
+	if n.asDecimal != nil {
+		return n.asDecimal.String()
+	}
+
+	return ""
+}
+
+func NumericFromString(s string) (*Numeric, error) {
+	if s == "" {
+		return nil, nil
+	}
+
+	// Check if the provided string contains a ".", because if it does not,
+	// the DecimalFromString will return it as int
+	split := strings.Split(s, ".")
+	if len(split) > 1 {
+		d, err := DecimalFromString(s)
+		if err != nil {
+			return nil, fmt.Errorf("error obtaining decimal from string: %s", err.Error())
+		}
+		return &Numeric{
+			asDecimal: &d,
+		}, nil
+	}
+
+	u, _ := UintFromString(s, 10)
+
+	return &Numeric{
+		asUint: u,
+	}, nil
+}
+
+// ScaleTo calculates the current contained value - decimal or uint - scaled to the target decimals.
+func (n *Numeric) ScaleTo(op, tdp int64) (*Uint, error) {
+	base := DecimalFromInt64(10)
+	if n.asDecimal != nil {
+		scaled := n.asDecimal.Mul(base.Pow(DecimalFromInt64(tdp)))
+		r, overflow := UintFromDecimal(scaled)
+		if overflow {
+			return nil, errors.New("failed to scale settlement data, overflow occurred")
+		}
+		return r, nil
+	}
+
+	if n.asUint != nil {
+		scaled := base.Pow(DecimalFromInt64(tdp - op))
+		r, overflow := UintFromDecimal(n.asUint.ToDecimal().Mul(scaled))
+		if overflow {
+			return nil, errors.New("failed to scale settlement data, overflow occurred")
+		}
+		return r, nil
+	}
+
+	return nil, nil
+}
+
+func (n *Numeric) SetUint(u *Uint) *Numeric {
+	n.asUint = u
+	n.asDecimal = nil
+	return n
+}
+
+func (n *Numeric) SetDecimal(d *Decimal) *Numeric {
+	n.asDecimal = d
+	n.asUint = nil
+
+	return n
+}
+
+func (n *Numeric) Decimal() *Decimal {
+	return n.asDecimal
```

Comment:
Same here: avoid sharing pointers

Reply:
I thought you said it should be sharing pointers - like here https://github.com/vegaprotocol/vega/pull/6935#discussion_r1048941709

Reply:
In that comment I meant that the implementation returned the underlying pointer, so the clone of the original object would share a pointer to the same data of the original.

Reply:
Understood, changed that to have a copy of the value and return pointer to it -> https://github.com/vegaprotocol/vega/pull/6935/commits/45c9f3f7de3c29edb1d92a134ee7e1e34f567730#diff-01e84bbc9c49dbdd4b9e45ec724423ea40a7c808577eebb58ed4f69f25ac4755R110-R124
Thanks!